test/unit_tests/arkode/CXX_serial/ark_test_adjoint_ark.cpp

```diff
@@ -0,0 +1,487 @@
+  const int ncheck                             = nsteps;
+  const sunbooleantype save_stages             = args.save_stages;
+  const sunbooleantype keep_check              = args.keep_checks;
+  SUNAdjointCheckpointScheme checkpoint_scheme = NULL;
+  SUNMemoryHelper mem_helper                   = SUNMemoryHelper_Sys(sunctx);
+  SUNAdjointCheckpointScheme_Create_Fixed(SUNDATAIOMODE_INMEM, mem_helper,
+                                          check_interval, ncheck, save_stages,
+                                          keep_check, sunctx, &checkpoint_scheme);
+  ARKodeSetAdjointCheckpointScheme(arkode_mem, checkpoint_scheme);
+
+  //
+  // Compute the forward solution
+  //
+
+  printf("\n-- Do forward problem --\n\n");
+
+  printf("Initial condition:\n");
+  N_VPrint(u);
+
+  forward_solution(sunctx, arkode_mem, checkpoint_scheme, t0, tf, dt, u);
+  if (check_forward_answer(u))
+  {
+    ++failcount;
+    fprintf(stderr,
+            ">>> FAILURE: forward solution does not match correct answer\n");
+  }
+  else { printf(">>> PASS\n"); }
+
+  //
+  // Create the adjoint stepper
+  //
+
+  printf("\n-- Do adjoint problem using Jacobian matrix --\n\n");
+
+  sunindextype num_params = 4;
+  N_Vector sensu0         = N_VClone(u);
+  N_Vector sensp          = N_VNew_Serial(num_params, sunctx);
+  N_Vector sens[2]        = {sensu0, sensp};
+  N_Vector sf             = N_VNew_ManyVector(2, sens, sunctx);
+
+  // Set the terminal condition for the adjoint system, which
+  // should be the the gradient of our cost function at tf.
+  dgdu(u, sensu0, params, tf);
+  dgdp(u, sensp, params, tf);
+
+  printf("Adjoint terminal condition:\n");
+  N_VPrint(sf);
+
+  SUNAdjointStepper adj_stepper;
+  ARKStepCreateAdjointStepper(arkode_mem, sf, &adj_stepper);
+
+  //
+  // Now compute the adjoint solution
+  //
+
+  SUNMatrix jac  = SUNDenseMatrix(neq, neq, sunctx);
+  SUNMatrix jacp = SUNDenseMatrix(neq, num_params, sunctx);
+
+  SUNAdjointStepper_SetJacFn(adj_stepper, ode_jac, jac, parameter_jacobian, jacp);
+
+  adjoint_solution(sunctx, adj_stepper, checkpoint_scheme, tf, t0, sf);
+  if (check_sensitivities(sf))
+  {
+    ++failcount;
+    fprintf(stderr,
+            ">>> FAILURE: adjoint solution does not match correct answer\n");
+  }
+  else { printf("\n>>> PASS\n"); }
+
+  //
+  // Now compute the adjoint solution using vjp
+  //
+
+  printf("\n-- Redo adjoint problem using VJP --\n\n");
+  if (!keep_check)
+  {
+    N_VConst(SUN_RCONST(1.0), u);
+    printf("Initial condition:\n");
+    N_VPrint(u);
+    ARKStepReInit(arkode_mem, ode_rhs, NULL, t0, u);
+    forward_solution(sunctx, arkode_mem, checkpoint_scheme, t0, tf, dt, u);
+    if (check_forward_answer(u))
+    {
+      ++failcount;
+      fprintf(stderr,
+              ">>> FAILURE: forward solution does not match correct answer\n");
+    };
+  }
+  dgdu(u, sensu0, params, tf);
+  dgdp(u, sensp, params, tf);
+  SUNAdjointStepper_ReInit(adj_stepper, u, t0, sf, tf);
```

Comment:
Should this be
```suggestion
  SUNAdjointStepper_ReInit(adj_stepper, u, tf);
```

Reply:
Im not following how the suggestion would work.